File: backend/app/schemas/pool.py
"""数据池相关 Pydantic 模型"""
from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, Field


class PoolItemCreate(BaseModel):
    """单条数据池记录写入"""
    pool_type: str = Field(..., description="plan|anomaly")
    source_system: str = "manual"
    source_ref: str | None = None
    title: str = Field(..., max_length=512)
    project_name: str | None = None
    person_name: str | None = None
    deadline: date | None = None
    description: str | None = None
    metric_type: str | None = None
    metric_value: float | None = None
    threshold: float | None = None
    deviation_pct: float | None = None
    raw_data: dict[str, Any] | None = None


class PoolItemUpdate(BaseModel):
    """编辑数据池记录（生成前修正）"""
    title: str | None = None
    project_name: str | None = None
    person_name: str | None = None
    deadline: date | None = None
    description: str | None = None
    metric_type: str | None = None
    metric_value: float | None = None
    threshold: float | None = None
    deviation_pct: float | None = None
    status: str | None = None
    skip_reason: str | None = None


class PoolItemOut(BaseModel):
    id: int
    pool_type: str
    source_system: str
    source_ref: str | None = None
    title: str
    project_name: str | None = None
    person_name: str | None = None
    deadline: date | None = None
    description: str | None = None
    metric_type: str | None = None
    metric_value: float | None = None
    threshold: float | None = None
    deviation_pct: float | None = None
    status: str
    work_order_id: int | None = None
    skip_reason: str | None = None
    raw_data: dict[str, Any] | None = None
    backfill_reason: str | None = None
    backfill_action: str | None = None
    backfilled_at: datetime | None = None
    created_at: datetime

    # 关联名称
    work_order_code: str | None = None

    model_config = {"from_attributes": True}


class PoolItemListOut(BaseModel):
    items: list[PoolItemOut]
    total: int
    page: int
    page_size: int


class GenerateRequest(BaseModel):
    pool_ids: list[int] = Field(..., min_length=1, max_length=500)


class GenerateResult(BaseModel):
    generated: int
    skipped: int
    errors: list[str]
    work_order_ids: list[int]


class PoolImportResult(BaseModel):
    imported: int
    skipped: int
    errors: list[str]


class BackfillRequest(BaseModel):
    reason: str | None = None
    action: str | None = None
    trigger_new_wo: bool = False
    new_wo_title: str | None = None
    new_wo_deadline: date | None = None
    new_wo_person_name: str | None = None


class BackfillOut(BaseModel):
    work_order_id: int
    reason: str | None = None
    action: str | None = None
    triggered_wo_id: int | None = None
    backfilled_at: datetime

File: frontend/src/api/pool.ts
import http from "./http";

export interface PoolItem {
  id: number;
  pool_type: string;
  source_system: string;
  source_ref?: string | null;
  title: string;
  project_name?: string | null;
  person_name?: string | null;
  deadline?: string | null;
  description?: string | null;
  metric_type?: string | null;
  metric_value?: number | null;
  threshold?: number | null;
  deviation_pct?: number | null;
  status: string;
  work_order_id?: number | null;
  skip_reason?: string | null;
  raw_data?: any;
  backfill_reason?: string | null;
  backfill_action?: string | null;
  backfilled_at?: string | null;
  created_at: string;
  work_order_code?: string | null;
}

export interface PoolListOut {
  items: PoolItem[];
  total: number;
  page: number;
  page_size: number;
}

export interface PoolImportResult {
  imported: number;
  skipped: number;
  errors: string[];
}

export interface GenerateResult {
  generated: number;
  skipped: number;
  errors: string[];
  work_order_ids: number[];
}

export interface BackfillResult {
  work_order_id: number;
  reason?: string | null;
  action?: string | null;
  triggered_wo_id?: number | null;
  triggered_wo_code?: string | null;
  backfilled_at: string;
  // 判断Agent 结果
  verdict?: string | null;
  judgment_reasoning?: string | null;
  judgment_suggestions?: {
    title?: string | null;
    deadline?: string | null;
    person_name?: string | null;
    priority?: string | null;
    action_adjustment?: string | null;
  } | null;
  judgment_confidence?: number | null;
}

// 数据池列表
export const listPoolItems = (params?: { pool_type?: string; status?: string; page?: number; page_size?: number }) =>
  http.get<any, PoolListOut>("/pool/items", { params });

// 单条详情
export const getPoolItem = (id: number) =>
  http.get<any, PoolItem>(`/pool/items/${id}`);

// 手动录入
export const createPoolItem = (data: Partial<PoolItem>) =>
  http.post<any, PoolItem>("/pool/items", data);

// 编辑
export const updatePoolItem = (id: number, data: Partial<PoolItem>) =>
  http.patch<any, PoolItem>(`/pool/items/${id}`, data);

// 删除
export const deletePoolItem = (id: number) =>
  http.delete(`/pool/items/${id}`);

// CSV 上传
export const uploadPoolCSV = (pool_type: string, file: File) => {
  const fd = new FormData();
  fd.append("file", file);
  return http.post<any, PoolImportResult>(`/pool/upload?pool_type=${pool_type}`, fd, {
    headers: { "Content-Type": "multipart/form-data" },
  });
};

// 批量生成工单
export const generateFromPool = (pool_ids: number[]) =>
  http.post<any, GenerateResult>("/pool/generate", { pool_ids });

// 一键生成全部
export const generateAllFromPool = (pool_type?: string) =>
  http.post<any, GenerateResult>("/pool/generate-all", null, { params: { pool_type } });

// AI表格同步
export const syncAITable = (base_id?: string, table_id?: string, pool_type?: string) =>
  http.post<any, PoolImportResult>("/pool/sync-aitable", null, { params: { base_id, table_id, pool_type } });

// 一键全链路同步：AITable→数据池→生成工单 + 钉盘「工单版」xlsx→工单
export interface FullSyncResult {
  aitable: { anomaly_synced: number; non_eam_synced: number };
  pool_generated: number;
  drive_imported: number;
  drive_files: number;
  errors: string[];
}
export const syncFull = () =>
  http.post<any, FullSyncResult>("/pool/sync-full", null);

// 回填
export const backfillWO = (wo_id: number, data: {
  reason?: string; action?: string; trigger_new_wo?: boolean;
  new_wo_title?: string; new_wo_deadline?: string; new_wo_person_name?: string;
  accept_judgment?: boolean; override_judgment?: boolean;
}) =>
  http.post<any, BackfillResult>(`/work-orders/${wo_id}/backfill`, data);

// 查看回填
export const getBackfill = (wo_id: number) =>
  http.get<any, BackfillResult>(`/work-orders/${wo_id}/backfill`);

// 判断Agent 导出
export const exportJudgment = (wo_id: number) =>
  http.get(`/work-orders/${wo_id}/export-judgment`, { responseType: "blob" });

// 判断Agent 导入
export const importJudgment = (wo_id: number, data: any) =>
  http.post<any, any>(`/work-orders/${wo_id}/import-judgment`, data);

// 人员看板
export const getPersonDashboard = (user_id: number) =>
  http.get<any, any>(`/dashboard/person/${user_id}`);

// 工单日历
export const getCalendar = (year: number, month: number, person_id?: number, project_id?: number) =>
  http.get<any, any>("/dashboard/calendar", { params: { year, month, person_id, project_id } });

// 趋势数据
export const getTrends = (months?: number) =>
  http.get<any, any>("/dashboard/trends", { params: { months } });